cognite_toolkit/_cdf_tk/resource_classes/capabilities.py

```diff
@@ -0,0 +1,507 @@
+import sys
+from types import MappingProxyType, UnionType
+from typing import Any, ClassVar, Literal, cast, get_args
+
+from pydantic import BaseModel, ModelWrapValidatorHandler, field_validator, model_serializer, model_validator
+from pydantic_core.core_schema import SerializerFunctionWrapHandler
+
+from cognite_toolkit._cdf_tk.utils.collection import humanize_collection
+
+if sys.version_info < (3, 11):
+    from typing_extensions import Self
+else:
+    from typing import Self
+
+from .base import BaseModelResource
+
+
+class Scope(BaseModelResource):
+    _scope_name: ClassVar[str]
+
+    @model_validator(mode="wrap")
+    @classmethod
+    def find_scope_cls(cls, data: Any, handler: ModelWrapValidatorHandler[Self]) -> Self:
+        if isinstance(data, Scope):
+            return cast(Self, data)
+        if not isinstance(data, dict):
+            raise ValueError(f"Invalid scope data '{type(data)}' expected dict")
+
+        if cls is not Scope:
+            return handler(data)
+        name, content = next(iter(data.items()))
+        if name not in _SCOPE_CLASS_BY_NAME:
+            raise ValueError(
+                f"invalid scope name '{name}'. Expected one of {humanize_collection(_SCOPE_CLASS_BY_NAME.keys(), bind_word='or')}"
+            )
+        cls_ = _SCOPE_CLASS_BY_NAME[name]
+        return cast(Self, cls_.model_validate(content))
+
+    @model_serializer(mode="wrap", when_used="always", return_type=dict)
+    def include_scope_name(self, handler: SerializerFunctionWrapHandler) -> dict:
+        if self._scope_name is None:
+            raise ValueError("Scope name is not set")
+        serialized_data = handler(self)
+        return {self._scope_name: serialized_data}
+
+
+class AllScope(Scope):
+    _scope_name = "all"
+
+
+class CurrentUserScope(Scope):
+    _scope_name = "currentuserscope"
+
+
+class IDScope(Scope):
+    _scope_name = "idScope"
+    ids: list[str]
+
+
+class IDScopeLowerCase(Scope):
+    """Necessary due to lack of API standardisation on scope name: 'idScope' VS 'idscope'"""
+
+    _scope_name = "idscope"
+    ids: list[str]
+
+
+class InstancesScope(Scope):
+    _scope_name = "instancesScope"
+    instances: list[str]
+
+
+class ExtractionPipelineScope(Scope):
+    _scope_name = "extractionPipelineScope"
+    ids: list[str]
+
+
+class PostgresGatewayUsersScope(Scope):
+    _scope_name = "usersScope"
+    usernames: list[str]
+
+
+class DataSetScope(Scope):
+    _scope_name = "datasetScope"
+    ids: list[str]
+
+
+class TableScope(Scope):
+    _scope_name = "tableScope"
+    dbs_to_tables: dict[str, list[str] | dict[Literal["tables"], list[str]]]
+
+
+class AssetRootIDScope(Scope):
+    _scope_name = "assetRootIdScope"
+    root_ids: list[str]
+
+
+class ExperimentsScope(Scope):
+    _scope_name = "experimentscope"
```

Comment:
Suggest sticking with singular 

```suggestion
class ExperimentScope(Scope):
    _scope_name = "experimentscope"
```